oak_client/src/attestation.rs

```diff
@@ -109,50 +134,109 @@ impl rustls::ServerCertVerifier for RemoteAttestationVerifier {
+                        "Couldn't convert public key to PEM: {}",
+                        error
+                    ));
+                    warn!("{:?}", error);
+                    error
+                })?;
+
+            // Extract X.509 TEE extension.
+            let extensions = parsed_certificate.extensions();
+            let tee_extension = extensions.get(&TEE_EXTENSION_OID).ok_or_else(|| {
+                rustls::TLSError::General("Couldn't find X.509 TEE extension".to_string())
             })?;
+            let tee_extension_value = parse_der(tee_extension.value)
+                .map_err(|error| {
+                    let error = rustls::TLSError::General(format!(
+                        "Couldn't parse X.509 TEE extension: {:?}",
+                        error
+                    ));
+                    warn!("{:?}", error);
+                    error
+                })?
+                .1
+                .as_str()
+                .map_err(|error| {
+                    let error = rustls::TLSError::General(format!(
+                        "Couldn't parse TEE report value: {:?}",
+                        error
+                    ));
+                    warn!("{:?}", error);
+                    error
+                })?;
 
-        // Use [`rustls::WebPKIVerifier`] to check certificate correctness.
-        // This check is required since [`rustls::DangerousClientConfig::set_certificate_verifier`]
-        // completely overwrites the default verifier.
-        // https://docs.rs/rustls/0.19.0/rustls/struct.DangerousClientConfig.html#impl
-        let verifier = rustls::WebPKIVerifier::new();
-        verifier.verify_server_cert(roots, certs, hostname, ocsp)
+            // Check that X.509 TEE extension contains a correct TEE report.
+            self.check_tee_report(tee_extension_value, &public_key)
+                .map_err(|error| {
+                    let error =
+                        rustls::TLSError::General(format!("Incorrect TEE report: {:?}", error));
+                    warn!("{:?}", error);
+                    error
+                })?;
+
+            // If root certificates were not provided - only accept self-signed certificates.
+            let root_certificates = if roots.is_empty() {
+                let mut root_certificates = rustls::RootCertStore::empty();
+                for certificate in certs.iter() {
+                    root_certificates.add(certificate).map_err(|error| {
+                        let error = rustls::TLSError::General(format!(
+                            "Incorrect self-signed certificate: {:?}",
+                            error
+                        ));
+                        warn!("{:?}", error);
+                        error
+                    })?;
+                }
+                root_certificates
+            } else {
+                roots.clone()
+            };
+
+            // Add certificate correctness check.
```

Comment:
Perhaps move this so that it is the first check? I don't think it matters for this case, but it is usually a good idea to do general validity checks before doing specific checks. In some cases doing specific checks before general validation allows attackers to gain extra information by sending invalid or malformed inputs.